mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.feature
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.api.java.JavaRDD
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.mllib.linalg.distributed.RowMatrix
+import org.apache.spark.rdd.RDD
+
+/**
+ * :: Experimental ::
+ * Transformer a vector use PCA
+ */
+
+object PCA {
+  /**
+   * Compute a Principal Components from RDD[Vector]
+   */
+  def create(k: Int, sources: RDD[Vector]): PCAModel = {
+    require(k > 1 && k <= sources.first().size)
+    
+    val mat = new RowMatrix(sources)
+    val pc = mat.computePrincipalComponents(k) match {
+      case pc: DenseMatrix =>
+        pc
+      /*
+       * Convert Sparse Vector to Dense Vector.
+       *
+       * Following code is for possible compatibility.
+       * RowMatrix.computePrincipalComponents is always returned a Dense Vector
+       */
+      case v =>
+        Matrices.dense(v.numRows, v.numCols, v.toArray)
+    }
+    new PCAModel(k, pc.asInstanceOf[DenseMatrix])
+  }
+
+  /**
+   * Compute a Principal Components from JavaRDD[Vector]
+   */
+  def create(k: Int, sources: JavaRDD[Vector]): PCAModel =
+    create(k, sources.rdd)
+}
+
+/**
+ * This class is wrapper of RowMatrix.computePrincipalComponents and Matrix.multiply
+ *
+ * @param k count of principal components.
+ * @param pc a principal components Matrix
+ */
+@Experimental
+class PCAModel private[mllib](val k: Int, val pc: DenseMatrix) extends VectorTransformer {
+  /**
+   * Transform a vector by computed Principal Components.
+   *
+   * @param vector vector to be transformed.
+   * @return transformed vector.
+   */
+  override def transform(vector: Vector): Vector = {
+    val mat = Matrices.dense(1, vector.size, vector.toArray).multiply(pc)
```

Comment:
`BLAS.gemv` should work. We might miss the implementation of dense matrix * sparse vector, but it should be easy to add.